### ISSUE TITLE
button_to generate incorrect hidden fields with params being an ActionController::Parameters

### ISSUE THREAD

**Issue by nanowang** [rails/rails#26802]
This issue is similar to #25256

# Steps to reproduce

1. assume `params` being:

```
<ActionController::Parameters {"query"=>"query", "controller"=>"posts", "action"=>"show"} permitted: false>
```

2.

```
<%= button_to 'name', '/path', {params: params} %>
```

# Expected behavior

the HTML should generate a structure like this:

```
<form>
  <some inputs here>
  <input type="hidden" name="query" value="query">
</form>
```

# Actual behavior

the HTML generates to this:

```
<form>
  <some inputs here>
  <input type="hidden" value="#&lt;ActionController::Parameters:0x007f9455615698&gt;" />
</form>
```

# Workaround

Using `params.to_h` instead of `params` works fine.

# System configuration

Rails version: 5.0.0.1

**Comment by BEaStia** [comment]
I'm afraid that passing unpermitted parameters is not the best way of work.

**Comment by maclover7** [comment]
Hi @nanowang! Unfortunately, this is not supported behavior in Rails. `ActionController::Parameters` does not respond to the `to_param` method, which is necessary when constructing the `button_to` helper. In order to fix this, I recommend creating a private `post_form_params` method or something along those lines where you can call `params.to_h` or return another object that adheres to the `to_param` contract. I've posted a small reproduction and solution script [here](https://gist.github.com/maclover7/1c7a3637dfd560b713e0a584873ca587) if you want to take a look at that. Thanks for opening up this issue!

**Comment by rafaelfranca** [comment]
Fixed in #26810
